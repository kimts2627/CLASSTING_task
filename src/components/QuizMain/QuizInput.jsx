import axios from "axios";
import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { setOptions, setQuiz } from "../../modules/quiz";

const QuizInput = () => {
  const [alertMsg, setMsg] = useState(false);
  const [categories, setCategories] = useState([]);
  const [currentOptions, setCurrentOptions] = useState({
    category: "",
    amount: "",
    difficulty: "",
  });
  const options = useSelector((state) => state.quiz.quizOptions);

  const dispatch = useDispatch();

  const onSetOptions = (newOptions) => {
    dispatch(setOptions(newOptions));
  };

  const onSetQuiz = (quiz) => {
    dispatch(setQuiz(quiz));
  };

  const handleChange = (e) => {
    let { name, value } = e.target;
    setCurrentOptions({ ...currentOptions, [name]: value });
  };

  const makeQuiz = async () => {
    if (Object.values(currentOptions).includes("")) {
      setMsg(true);
      return;
    }
    onSetOptions(currentOptions);
    const res = await axios(
      `https://opentdb.com/api.php?amount=${currentOptions.amount}&category=${currentOptions.category}&difficulty=${currentOptions.difficulty}&type=multiple`
    );
    const data = res.data.results;
    onSetQuiz(data);
  };

  useEffect(async () => {
    const res = await axios("https://opentdb.com/api_category.php");
    const data = res.data.trivia_categories;
    setCategories(data);
  }, []);

  return (
    <section className="bg-white w-full h-96">
      <form className="flex flex-col items-center jutify-evenly">
        <label>Quiz Options</label>
        <select name="category" className="w-44 border" onChange={(e) => handleChange(e)}>
          <option value="">--Category--</option>
          {categories.map((category) => (
            <option key={category.id} value={category.id}>
              {category.name}
            </option>
          ))}
        </select>
        <select name="amount" className="w-44 border" onChange={(e) => handleChange(e)}>
          <option value="">--Amount--</option>
          {["5", "10"].map((num) => (
            <option key={num} value={num}>
              {num}
            </option>
          ))}
        </select>
        <select name="difficulty" className="w-44 border" onChange={(e) => handleChange(e)}>
          <option value="">--Difficulty--</option>
          {["easy", "medium", "hard"].map((diff) => (
            <option key={diff} value={diff}>
              {diff}
            </option>
          ))}
        </select>
        {alertMsg ? <p className="text-xs text-red-500">please fill all options</p> : null}
        <input type="button" value="Start Quiz!" onClick={makeQuiz} />
      </form>
    </section>
  );
};

export default QuizInput;
