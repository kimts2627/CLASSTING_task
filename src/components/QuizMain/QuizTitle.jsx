import React from "react";

const QuizTitle = () => {
  return (
    <header className="Kanit text-5xl text-white flex flex-col items-center">
      <span className="text-yellow-400">
        Random <span className="text-white">Quiz</span>
      </span>
      <p className="text-lg">Select your quiz type! and enjoy!</p>
    </header>
  );
};

export default QuizTitle;
