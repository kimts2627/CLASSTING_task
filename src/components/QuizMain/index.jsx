import React from "react";
import QuizInput from "./QuizInput";
import QuizTitle from "./QuizTitle";

const QuizMain = () => {
  return (
    <main className="absolute flex flex-col w-full h-full items-center justify-evenly">
      <QuizTitle />
      <QuizInput />
    </main>
  );
};

export default QuizMain;
