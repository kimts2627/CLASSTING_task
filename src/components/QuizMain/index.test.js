import React from "react";
import { mount, configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import QuizMain from "./index";
import QuizInput from "./QuizInput";
import QuizTitle from "./QuizTitle";
import { store } from "../../index";
import { Provider } from "react-redux";

configure({ adapter: new Adapter() });
let component = null;

describe("QuizMain", () => {
  describe("QuizMain", () => {
    it("QuizMain 컴포넌트가 정상적으로 렌더링 되어야 함", () => {
      component = mount(
        <Provider store={store}>
          <QuizMain />
        </Provider>
      );
      expect(component.length).toBe(1);
    });
  });

  describe("QuizInput", () => {
    it("QuizInput 컴포넌트가 정상적으로 렌더링 되어야 함", () => {
      component = mount(
        <Provider store={store}>
          <QuizInput />
        </Provider>
      );
      expect(component.length).toBe(1);
    });

    it("select tag들이 정상적으로 렌더링 되어야 함", () => {
      expect(component.find("select").length).toEqual(3);
    });

    it("button input tag가 정상적으로 렌더링 되어야 함", () => {
      expect(component.find("input").length).toEqual(1);
    });
  });

  describe("QuizTitle", () => {
    it("QuizTitle 컴포넌트가 정상적으로 렌더링 되어야 함", () => {
      component = mount(
        <Provider store={store}>
          <QuizTitle />
        </Provider>
      );
      expect(component.length).toBe(1);
    });
  });

  it("스냅샷과 결과물이 일치해야 함", () => {
    expect(component).toMatchSnapshot();
  });
});
