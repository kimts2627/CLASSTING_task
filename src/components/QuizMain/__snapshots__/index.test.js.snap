// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizMain 스냅샷과 결과물이 일치해야 함 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <QuizTitle>
    <header
      className="Kanit text-6xl text-white flex flex-col items-center"
    >
      <span
        className="text-yellow-400"
      >
        Random 
        <span
          className="text-white"
        >
          Quiz
        </span>
      </span>
      <p
        className="text-lg"
      >
        Select your quiz type! and enjoy!
      </p>
    </header>
  </QuizTitle>
</Provider>
`;
